### PR TITLE
Rawsong to song

### PR DESCRIPTION
**Pull request by heus92**
Django custom command made to transform scrapped `GeniusRawSong` objects to `Track` objects with correct formatting and relevant information. 

Works the same way the `rawartist_to_artist` command works meaning
- if you want to **create** new objects from data that has not been imported yet use the following command
```
python manage.py --update False
```

- if you want to **update** the existing objects from newly imported data use the following command
```
python manage.py --update True
```

*NB: would be awesome if commands could be tested on a dev instance of our online db before merging*

### CODE REVIEW

File: api/models.py [hunk 1]
```diff
@@ -71,12 +71,36 @@ class Track(models.Model):
     """
     Clean track representation based on data extracted from the GeniusRawSong table
     """
-    genius_annotation_count = models.IntegerField(null=True)
     genius_id = models.IntegerField(unique=True, null=True)
-    recording_location = models.CharField(max_length=300, null=True)
-    release_date = models.DateTimeField(null=True)
-    title = models.CharField(max_length=300, null=False)
+
+    title = models.CharField(max_length=300, null=True)
```
Comment (LouisonGitzinger) [inline api/models.py]: this field should never be `null` I guess

File: api/models.py [hunk 2]
```diff
@@ -71,12 +71,36 @@ class Track(models.Model):
     """
     Clean track representation based on data extracted from the GeniusRawSong table
     """
-    genius_annotation_count = models.IntegerField(null=True)
     genius_id = models.IntegerField(unique=True, null=True)
```
Comment (LouisonGitzinger) [inline api/models.py]: this can't be null

File: api/models.py [hunk 3]
```diff
@@ -71,12 +71,36 @@ class Track(models.Model):
...
-    recording_location = models.CharField(max_length=300, null=True)
-    release_date = models.DateTimeField(null=True)
-    title = models.CharField(max_length=300, null=False)
+
+    title = models.CharField(max_length=300, null=True)
+    full_title = models.CharField(max_length=300, null=True)
+    lyrics_state = models.CharField(max_length=15, null=True)
+    release_date = models.CharField(max_length=10, null=True)
     genius_url = models.TextField(validators=[URLValidator()], null=True)
```
Comment (LouisonGitzinger) [inline api/models.py]: This can't be null

File: api/models.py [hunk 4]
```diff
@@ -71,12 +71,36 @@ class Track(models.Model):
...
-    genius_annotation_count = models.IntegerField(null=True)
     genius_id = models.IntegerField(unique=True, null=True)
-    recording_location = models.CharField(max_length=300, null=True)
-    release_date = models.DateTimeField(null=True)
-    title = models.CharField(max_length=300, null=False)
+
+    title = models.CharField(max_length=300, null=True)
+    full_title = models.CharField(max_length=300, null=True)
+    lyrics_state = models.CharField(max_length=15, null=True)
+    release_date = models.CharField(max_length=10, null=True)
```
Comment (LouisonGitzinger) [inline api/models.py]: This should be converted as a timestamp and stored in a DateTimeField (see GeniusRawArtist model for example)

File: api/models.py [hunk 5]
```diff
@@ -71,12 +71,36 @@ class Track(models.Model):
...
     genius_url = models.TextField(validators=[URLValidator()], null=True)
+    description = models.TextField(null=True)
+
+    album_id = models.IntegerField(null=True)
```
Comment (LouisonGitzinger) [inline api/models.py]: Here, we definitely need to create an Album model, with a foreign key. The foreign key must be the album id and not the album genius_id. It means that for each song, we need to check if it belongs to an album and if so, we need to find the album id using its genius_id or create the album.

Comment (LouisonGitzinger) [reply api/models.py]: Open an issue here

File: api/models.py [hunk 6]
```diff
@@ -71,12 +71,36 @@ class Track(models.Model):
...
+    primary_artist_id = models.IntegerField(null=True)
```
Comment (LouisonGitzinger) [inline api/models.py]: This can be null ?

File: api/models.py [hunk 7]
```diff
@@ -71,12 +71,36 @@ class Track(models.Model):
...
+    genius_unreviewed_annotations_count = models.IntegerField(null=True)
+    genius_contributors_count = models.IntegerField(null=True)
+    recording_location = models.CharField(max_length=300, null=True)
+
+    lyrics = models.TextField(null=True)
+
+    added = models.DateTimeField(auto_now_add=True)
+    updated = models.DateTimeField(auto_now=True)
+
     # -------- Future release keys --------
```
Comment (LouisonGitzinger) [inline api/models.py]: Clean up these comments if not useful anymore

File: scrapper/management/commands/rawsong_to_song.py [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+        if options['is_update']:
+            update_songs()
+        else:
+            unhandled_songs = get_unhandled_songs()
+            for unhandled_song in unhandled_songs:
+                data = json.loads(unhandled_song.data)
+                genius_song = data['response']['song']
+
+                genius_song_description = json_html_to_plain_text(genius_song['description']['dom']['children'])
+                if genius_song_description == "?":
```
Comment (LouisonGitzinger) [inline scrapper/management/commands/rawsong_to_song.py]: What is the difference between a null description and a description filled with "?" in the raw data ?
Do we clearly take the decision to make no difference between "?" and None ?

File: scrapper/management/commands/rawsong_to_song.py [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+                    genius_song_description = None
+
+                album_id = None
+                if genius_song['album'] is not None:
```
Comment (LouisonGitzinger) [inline scrapper/management/commands/rawsong_to_song.py]:
```python
if not genius_song['album']:
```
works the same and less redundant

Comment (LouisonGitzinger) [reply scrapper/management/commands/rawsong_to_song.py]: Instead of : 
```python
album_id = None
if genius_song['album'] is not None:
album_id = genius_song['album']['id']
```
I would event write : 
```python
album_id = genius_song['album']['id'] if genius_song['album'] else None
```

File: api/models.py [hunk 10]
```diff
@@ -71,12 +71,36 @@ class Track(models.Model):
...
+    title = models.CharField(max_length=300, null=True)
+    full_title = models.CharField(max_length=300, null=True)
+    lyrics_state = models.CharField(max_length=15, null=True)
+    release_date = models.CharField(max_length=10, null=True)
     genius_url = models.TextField(validators=[URLValidator()], null=True)
+    description = models.TextField(null=True)
+
+    album_id = models.IntegerField(null=True)
+    primary_artist_id = models.IntegerField(null=True)
+    secondary_artists_ids = ArrayField(models.IntegerField(), null=True)
```
Comment (LouisonGitzinger) [inline api/models.py]: this is not clear, we need to know where these ids point to (Genius here of course). 
If we do the same with Spotify afterwards, we have to specify the id origin

File: scrapper/management/commands/rawsong_to_song.py [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+                if genius_song['media']:
+                    for media_source in genius_song['media']:
+                        if media_source['provider'] == 'youtube':
+                            genius_song_media_youtube = media_source['url']
+                        if media_source['provider'] == 'spotify':
+                            genius_song_media_spotify = media_source['url']
+                        if media_source['provider'] == 'apple_music':
+                            genius_song_media_apple_music = media_source['url']
+                        if media_source['provider'] == 'soundcloud':
+                            genius_song_media_soundcloud = media_source['url']
```
Comment (LouisonGitzinger) [inline scrapper/management/commands/rawsong_to_song.py]: This part is ok. To improve it, it is possible to handle it with a list of media like : 
`medias = ['youtube', 'spotify', 'apple_music', soundcloud']` and iterate over it. 
Just in case Genius adds a new media later.

File: scrapper/management/commands/rawsong_to_song.py [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+                    title = genius_song['title'],
+                    full_title = genius_song['full_title'],
+                    lyrics_state = genius_song['lyrics_state'],
+                    release_date = genius_song['release_date'],
+                    genius_url = genius_song['url'],
+                    description = genius_song_description,
+
+                    album_id = album_id,
+                    primary_artist_id = genius_song['primary_artist']['id'],
+                    secondary_artists_ids = featured_artists_ids,
```
Comment (LouisonGitzinger) [inline scrapper/management/commands/rawsong_to_song.py]: specify where the ids come from (see comment in models.py)

File: scrapper/management/commands/rawsong_to_song.py [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+                    genius_verified_annotations_count = genius_song['stats']['verified_annotations'],
+                    genius_unreviewed_annotations_count = genius_song['stats']['unreviewed_annotations'],
+                    genius_contributors_count = genius_song['stats']['contributors'],
+                    lyrics = unhandled_song.lyrics
+                )
+
+
+
+
+def update_songs():
```
Comment (LouisonGitzinger) [inline scrapper/management/commands/rawsong_to_song.py]: This part is redundant, prefer an empty function here and leave a comment to say that in case we want to update the model, we should write the update code in this function

File: api/models.py (hunk 1 above)
Comment (heus92) [reply api/models.py]: 👍

File: api/models.py (hunk 2 above)
Comment (heus92) [reply api/models.py]: 👍

File: api/models.py (hunk 3 above)
Comment (heus92) [reply api/models.py]: 👍

File: api/models.py (hunk 6 above)
Comment (heus92) [reply api/models.py]: as it's supposed to reference the primary artist genius_id I guess not since the source is genius. Removing `null=True` 👍

File: api/models.py (hunk 4 above)
Comment (heus92) [reply api/models.py]: I'd rather user a simple DateField 👍

**Comment by heus92** [comment]
@LouisonGitzinger changes ✅ 
Review <3

File: api/models.py [hunk 14]
```diff
@@ -71,26 +71,26 @@ class Track(models.Model):
     """
     Clean track representation based on data extracted from the GeniusRawSong table
     """
-    genius_id = models.IntegerField(unique=True, null=True)
 
-    title = models.CharField(max_length=300, null=True)
-    full_title = models.CharField(max_length=300, null=True)
+    title = models.CharField(max_length=255)
```
Comment (LouisonGitzinger) [inline api/models.py]: Why did you change for 255 ? Why did you remove the null=True ?

File: api/models.py [hunk 15]
```diff
@@ -71,26 +71,26 @@ class Track(models.Model):
...
+    full_title = models.TextField()
```
Comment (LouisonGitzinger) [inline api/models.py]: Why text field ? I think charfield is more appropriate here because full title will never be an entire text

File: api/models.py [hunk 16]
```diff
@@ -71,26 +71,26 @@ class Track(models.Model):
...
     lyrics_state = models.CharField(max_length=15, null=True)
-    release_date = models.CharField(max_length=10, null=True)
-    genius_url = models.TextField(validators=[URLValidator()], null=True)
-    description = models.TextField(null=True)
+    release_date = models.DateField(auto_now=False, auto_now_add=False, null=True)
+    genius_url = models.TextField(validators=[URLValidator()])
```
Comment (LouisonGitzinger) [inline api/models.py]: Plz give a value to the `null` parameter, even if it is false

File: api/models.py [hunk 17]
```diff
@@ -71,26 +71,26 @@ class Track(models.Model):
...
+    secondary_artists_genius_ids = ArrayField(models.IntegerField())
 
     youtube_url = models.TextField(validators=[URLValidator()], null=True)
     spotify_url = models.TextField(validators=[URLValidator()], null=True)
     apple_music_url = models.TextField(validators=[URLValidator()], null=True)
     soundcloud_url = models.TextField(validators=[URLValidator()], null=True)
 
     genius_is_hot = models.BooleanField(default=False)
-    genius_annotation_count = models.IntegerField(null=True)
+    genius_annotation_count = models.IntegerField()
```
Comment (LouisonGitzinger) [inline api/models.py]: null param value

File: scrapper/management/commands/rawsong_to_song.py (hunk 8 above)
Comment (LouisonGitzinger) [reply scrapper/management/commands/rawsong_to_song.py]: Why not directly return a `None` value from the `json_html_to_plain_text` in this case ?

File: api/models.py (hunk 14 above)
Comment (heus92) [reply api/models.py]: - changed to 255 because that's [Django maximum possible value](https://docs.djangoproject.com/en/2.0/ref/databases/)
> Any fields that are stored with VARCHAR column types have their max_length restricted to 255 characters if you are using unique=True for the field. This affects CharField, SlugField.

File: api/models.py (hunk 15 above)
Comment (heus92) [reply api/models.py]: - when running the `rawsong_to_song` command the script encountered errors on some titles which were longer than 255 characters. Thus `TextField` to remove all those errors
- I removed the `null = True` for the same reasons I removed it in the `title` field

File: api/models.py (hunk 17 above)
Comment (heus92) [reply api/models.py]: this should never be null no ? 🤷‍♂️

File: scrapper/management/commands/rawsong_to_song.py [hunk 18]
```diff
@@ -127,12 +127,18 @@ def get_genius_song_medias(genius_song):
     """
     song_medias = {}
     medias_list = ['youtube', 'spotify', 'apple_music', 'soundcloud']
-    for genius_song_media in genius_song['media']:
+
+    if genius_song['media']:
+        for genius_song_media in genius_song['media']:
+            for media_source in medias_list:
```
Comment (LouisonGitzinger) [inline scrapper/management/commands/rawsong_to_song.py]: double for loop can be avoided here, try : 
```python
if 'media' in genius_song:
     for genius_song_media in genius_song['media']:
          if genius_song_media['provider']: # check if the provider key exists
               if genius_song_media['provider'] in medias_list:
                    song_medias[media_source] = genius_song_media['url']
               else:
                    song_medias[media_source] = None
…
               
```
